### DISCUSSION TITLE
Why use setup-python action instead of a python container?

### DISCUSSION THREAD

**Discussion by tetienne** [community/community#25075] [Actions]
<p>Can you explain the pros and cons of both solutions?</p>
<pre><code>steps:
-uses:actions/checkout@v1-uses:actions/setup-python@v1with:
python-version:'3.7'-run:python my\_script.py  
</code></pre>
<p>vs</p>
<pre><code>container: python:3.7
runs-on: ubuntu-latest
steps:
    - uses: actions/checkout@v1
    - run: python my_script.py
</code></pre>

**Comment by BrightRan** [comment]
<p>I have a few summary here for your reference:</p>
<ul>
<li>If using <a href="https://github.com/marketplace/actions/setup-python-environment" rel="nofollow noopener">setup-python action</a> in a job which does not set a container, the setup-python step will run in its container, other steps that don’t already specify a container will directly run on the host specified by <strong>runs-on.</strong>  And other container actions will run in themselves container.</li>
<li>If set a container on the job, any steps in this job that don’t already specify a container will run in this container. Other container actions in this job will run as sibling containers on the same network with the same volume mounts.</li>
<li>The setup-python action only uses versions of Python already installed in the cache, and the action will fail if no matching versions are found. More details about the available versions of Python that are installed see <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/software-installed-on-github-hosted-runners" rel="nofollow noopener">here</a>. Current available versions of Python are listed below. If set a python container on the job, you have more versions of Python to select.</li>
</ul>

**Comment by tetienne** [comment]
<p>Thank you for your detailed answer. But I think you miss the main point of my question. I will rephrase.</p>
<p>My job needs Python 3.7 to run. Today I have two solutions: actions/setup-python or container: python:3.7.</p>
<p>In both cases, my job will pass.</p>
<p>What’s the pro and cons of these two solutions?</p>

**Comment by chrispat** [comment, marked as answer]
<p><a class="mention" href="/tetienne">@tetienne</a> Containers are only an option on Linux.  If you need to test your python application on Windows or MacOS then using setup-python is the right option.  It can also be a lot faster as in many cases you can use a python that is already downloaded on the runner rather than having to wait for a container to download.</p>

**Comment by tetienne** [comment]
<p>Oh I was not aware of container restricted only to Linux host.</p>
<p>I will try to perform some tests to see the speed difference between both. But indeed it makes sense.</p>
<p>Thx for your answer.</p>
